### PR TITLE
PERF: use uniqueness_check from monotonic check when possible

### LINKED ISSUE #14266: Large Monotonic Index Objects Always Allocate Hash Tables on get_loc
Historically, large monotonically-increasing `Index` objects would attempt to avoid creating a large hash table on `get_loc` calls. In service of this goal, many `IndexEngine` methods have guards like the one in `DatetimeEngine.get_loc`:

```python
        if self.over_size_threshold and self.is_monotonic_increasing:
            if not self.is_unique:
                val = _to_i8(val)
                return self._get_loc_duplicates(val)
            # Do lookup using `searchsorted`
        # Do lookup using hash table.
```

Since at least 5eecdb2dd94719e1fd097ce3fb046697445a3d7f, `self.is_unique` has been implemented as a property that would force a hash table to be created **unless the index had already been marked as unique**.  Until https://github.com/pydata/pandas/pull/10199, the `is_monotonic_increasing` property would perform a check that would [sometimes set `self.unique` to True](https://github.com/pydata/pandas/pull/10199/files#diff-4f3357a9f53e943087e2778134494905L236), which would prevent the large hash table allocation.  After the commit linked above, however, the only code path that ever sets `IndexEngine.unique` is in [`IndexEngine.initialize`](https://github.com/pydata/pandas/blob/a7469cf98275a183ad2e4bfafa9706a1ef8d035e/pandas/index.pyx#L269-L270), which [unconditionally creates a hash table before setting the unique flag.](https://github.com/pydata/pandas/blob/a7469cf98275a183ad2e4bfafa9706a1ef8d035e/pandas/index.pyx#L266-L267).

#### Code Sample, a copy-pastable example if possible

```python
import os
import humanize
import psutil
import pandas as pd


def get_mem_usage():
    pid = os.getpid()
    proc = psutil.Process(pid)
    return humanize.naturalsize(proc.memory_full_info().uss)

print("Pandas Version: " + pd.__version__)
print("Before Index Creation: " + get_mem_usage())

# The cutoff for allocating a hash table inside the index is supposed to be
# 1,000,000 entries in the index.  This index is about 10x larger.
data = pd.date_range('1990', '2016', freq='min')

print("After Index Creation: " + get_mem_usage())

# Trigger a hash of the index's contents.
data.get_loc(data[5])

print("After get_loc() call: " + get_mem_usage())
```

#### Output (Old Pandas):
```
$ python repro.py
Pandas Version: 0.16.1
Before Index Creation: 36.8 MB
After Index Creation: 146.5 MB
After get_loc() call: 146.6 MB
```

#### Output (Pandas 0.18.1)
```
$ python repro.py
Pandas Version: 0.18.1
Before Index Creation: 47.6 MB
After Index Creation: 157.4 MB
After get_loc() call: 698.7 MB
```

For some context, I found this after the internal Jenkins build for Zipline (which makes heavy use of large minutely `DatetimeIndex`es to represent trading calendars) started failing with memory errors after merging https://github.com/quantopian/zipline/pull/1339.

Assuming that the memory-saving behavior of older pandas is still desired, I think the right immediate fix for this is to change `IndexEngine._do_unique_check` to actually do a uniqueness check instead of just forcing a hash table creation.  Reading through the code, however, there are a bunch of ways that large `Index`es could still hit code paths that trigger hash table allocations.  For example, `DatetimeEngine.__contains__` guards against `self.over_size_threshold`, but none of the other `IndexEngine` subclasses do.  A more significant refactor is probably needed to provide a meaningful guarantee that indices don't consume too much memory.

#### output of ``pd.show_versions()``

<details>
In [4]: pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.10.final.0
python-bits: 64
OS: Linux
OS-release: 4.2.0-16-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.16.1
nose: 1.3.7
Cython: 0.22.1
numpy: 1.9.2
scipy: 0.15.1
statsmodels: 0.6.1
IPython: 5.1.0
sphinx: 1.3.4
patsy: 0.4.0
dateutil: 2.4.2
pytz: 2015.4
bottleneck: 1.0.0
tables: None
numexpr: 2.4.6
matplotlib: 1.4.3
openpyxl: None
xlrd: 0.9.4
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.8
pymysql: None
psycopg2: 2.6.1 (dt dec pq3 ext lo64)
</details>

### PR DESCRIPTION
**Pull request by jreback**
closes #14266

### CODE REVIEW

**Comment by jreback** [comment]
@ssanderson @llllllllll 

can you give this a check (ideally you can perf test this as well)......

**Comment by TomAugspurger** [comment]
Is this big enough to warrant a second release candidate? (you have a better understanding for how big a change this is).

**Comment by jreback** [comment]
@TomAugspurger well our entire test suite passes, so I don't think this is a big deal as not changing  functionaility (this is almost a bug fix).

File: doc/source/whatsnew/v0.19.0.txt [hunk 1]
```diff
@@ -1404,6 +1404,7 @@ Performance Improvements
 - Improved performance of datetime string parsing in ``DatetimeIndex`` (:issue:`13692`)
 - Improved performance of hashing ``Period`` (:issue:`12817`)
 - Improved performance of ``factorize`` of datetime with timezone (:issue:`13750`)
+- Improved performance of of not eagerly creating indexing hashtables on larges Indexes (:issue:`14266`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.0.txt]: sp, I know!

**Comment by jreback** [comment]
some nice perf improvements as well (note these are backwards, in that a7469rc9 is master)
```
  [9ced5d50] [a7469cf9]
+   14.98μs    54.75ms   3656.26  indexing.series_getitem_scalar.time_series_getitem_scalar
+   30.83μs    54.60ms   1770.70  indexing.series_ix_scalar.time_series_ix_scalar
+   50.48μs    52.97ms   1049.33  indexing.series_loc_slice.time_series_loc_slice
+   54.64μs    53.11ms    972.14  indexing.series_ix_slice.time_series_ix_slice
+    2.21ms    56.38ms     25.52  timeseries.timeseries_large_lookup_value.time_timeseries_large_lookup_value
+    1.36ms     4.17ms      3.08  timeseries.timeseries_datetimeindex_offset_delta.time_timeseries_datetimeindex_offset_delta
+  231.80μs   514.32μs      2.22  timeseries.datetimeindex_add_offset.time_datetimeindex_add_offset
-    3.45ms     1.50ms      0.43  index_object.index_datetime_set_difference.time_index_datetime_difference_disjoint
-   51.35ms    21.71ms      0.42  frame_methods.frame_getitem_single_column.time_frame_getitem_single_column
-  312.45μs   129.48μs      0.41  timeseries.timeseries_asof_single.time_timeseries_asof_single
-  239.91μs    92.65μs      0.39  index_object.datetime_index_union.time_datetime_index_union
SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

**Comment by jreback** [comment]
@ssanderson if you could test out would be great.

**Comment by ssanderson** [comment]
@jreback I'll take a look shortly.

**Comment by llllllllll** [comment]
This change looks good to me, not sure if I can easily test zipline with this though. Running the memory usage example from the issue shows the memory usage pretty constant across the get_loc call. Thanks for the really quick fix!

**Comment by ssanderson** [comment]
I'm installing this into a fresh zipline build to see if it runs (there may
be upgrade work that needs to happen to be compat with 0.19, so if that
doesn't work I'll probably just confirm the minimal repro case as well.)

**Comment by ssanderson** [comment]
LGTM.  Thanks for the quick turnaround @jreback.

**Pull request closed without merging**